Switch to structured logging

pinikomarov opened the pull request:
This automatically adds additional Log fields like reconcile_id etc.. from the controller context.

Follow up from: https://github.com/openstack-k8s-operators/keystone-operator/pull/220
Epic :  [OSP-22582](https://issues.redhat.com/browse/OSP-22582) Switch Operators to structured logging

before :
```bash
{"level":"info","ts":"2023-07-04T10:11:14.655+0300","logger":"controllers.OpenStackControlPlane","msg":"Reconciling Glance","Glance.Namespace":"openstack","Glance.Name":"glance"}
```

after:
```bash
2023-10-15T02:37:10.941+0300    INFO    Controllers.OpenStackClient     OpenStackClient CR values       {"controller": "openstackclient", "controllerGroup": "client.openstack.org", "controllerKind": "OpenStackClient", "OpenStackClient": {"name":"openstackclient","namespace":"openstack"}, "namespace": "openstack", "name": "openstackclient", "reconcileID": "21a5c4af-ea9a-4c5c-aae2-78c77b76b01c", "Name": "openstackclient", "Namespace": "openstack", "Secret": "openstack-config-secret", "Image": "quay.io/podified-antelope-centos9/openstack-openstackclient@sha256:66a098b20c2a26ffc43b7d21f76b17bba386c3eebcc33ed48fb31208896d9761"}
```

set log dev true for console output

pinikomarov commented:
Hi @gibizer,  can you take a look ? 

I'm running this on titan88 using install yamls with the patch I'm getting : 
```
/root/openstack-operator/controllers/core/openstackcontrolplane_controller.go:38:2: no such package located
Error: not all generators ran successfully
```

pinikomarov commented:
OH nevermind , I got 

-: import cycle not allowed: import stack: [github.com/openstack-k8s-operators/openstack-operator github.com/openstack-k8s-operators/openstack-operator/controllers/core github.com/openstack-k8s-operators/openstack-operator/pkg/openstack github.com/openstack-k8s-operators/openstack-operator/controllers/core]

so it's denying import of the  for the owned resources I guessing ..
  [< 1 minute ago](https://redhat-internal.slack.com/archives/CQXJFGMK6/p1688028270555299?thread_ts=1687904614.038659&cid=CQXJFGMK6)
so openstackcontrolplane_controller.go which is importing "[github.com/openstack-k8s-operators/openstack-operator/pkg/openstack](http://github.com/openstack-k8s-operators/openstack-operator/pkg/openstack)"
is calling pkg/openstack/ovn.go which is importing openstackcontrolplane_controller.go  for the GetLog func.
and that in turn causes re-import of the ["openstack-operator/pkg/openstack](http://github.com/openstack-k8s-operators/openstack-operator/pkg/openstack)" module. that's the import cycle go was talking about

gibizer commented:
It seems openstack-operator uses a different log format compared to keystone-operator.
keystone-operator:
```
2023-07-04T10:58:40Z	INFO	Controllers.KeystoneAPI	Reconciling Service	{"controller": "keystoneapi", "controllerGroup": "keystone.openstack.org", "controllerKind": "KeystoneAPI", "KeystoneAPI": {"name":"keystone","namespace":"openstack"}, "namespace": "openstack", "name": "keystone", "reconcileID": "ecd75bab-0441-4d17-9342-91416beb6d15"}
```
openstack-operator:
```
{"level":"info","ts":"2023-07-04T10:17:44.052+0300","logger":"Controllers.OpenstackControlPlane","msg":"Setting StatefulSet","controller":"openstackcontrolplane","controllerGroup":"core.openstack.org","controllerKind":"OpenStackControlPlane","OpenStackControlPlane":{"name":"openstack-network-isolation","namespace":"openstack"},"namespace":"openstack","name":"openstack-network-isolation","reconcileID":"9c199ee4-2e92-46ad-943c-fc977b70b9f5"}
```

I guess the difference is in the logger setup:
https://github.com/openstack-k8s-operators/openstack-operator/blob/50667049a677e32bcc846c694fea95b83bde5ee2/main.go#L111-L122

https://github.com/openstack-k8s-operators/keystone-operator/blob/4371f9602724378ac0c39eff7e635035afc761a3/main.go#L69-L75

The dev mode difference is added in https://github.com/openstack-k8s-operators/openstack-operator/commit/6435a06ef2017fa79799782551fe0c039ac1b991 But that commit does not have much reasoning why we are disabled dev mode by default. Maybe @fao89 or @abays knows more.

fao89 commented:
I didn't have a special reason for setting it to `false` as default, I just wanted to make the `development` configurable, no objections for making the `true` the default

pinikomarov commented:
thanks @fao89 I changed it in the patch
@gibizer I updated the zap logging option to dev true so it's console output mode.
I also updated the commit message at the top to reflect that

gibizer reviewed (changes requested):
I have two small requests inline.

gibizer commented:
> I didn't have a special reason for setting it to `false` as default, I just wanted to make the `development` configurable, no objections for making `true` the default

OK. So this is something we might want to decided together and sync in each operator. I don't know the actual implication of the devMode logger versus the non devMode one. Other then that actually log line format it emits. The devMode one is easier for humans to parse, but *maybe* the devMode:false one is better from machine processing?

pinikomarov commented:
> > I didn't have a special reason for setting it to `false` as default, I just wanted to make the `development` configurable, no objections for making `true` the default
> 
> OK. So this is something we might want to decided together and sync in each operator. I don't know the actual implication of the devMode logger versus the non devMode one. Other then that actually log line format it emits. The devMode one is easier for humans to parse, but _maybe_ the devMode:false one is better from machine processing?

I think as far as usage goes , the log format can always be changed by the DEV_MODE env, so anyone needing this can simply use that for logging with json.
I believe the majority of reviewers and debuggers right now are human so I think the console (dev true) mode is ok for now.
It is a point I will raise when I ask for reviews with a broader audience though.

pinikomarov commented:
/retest-required

pinikomarov commented:
/retest

pinikomarov commented:
/retest-required

pinikomarov commented:
seeing some unrealeted edpm/nmstate failures in CI : 
https://review.rdoproject.org/zuul/build/ab733d28c3584ea2b1b0493eb86816bc:
`nError from server (InternalError): error when creating \"/tmp/tmp.s9NWVJVmn4/tmp.C0br6ep3ph\": Internal error occurred: failed calling webhook \"nodenetworkconfigurationpolicies-mutate.nmstate.io\": failed to call webhook: Post \"https://nmstate-webhook.openshift-nmstate.svc:443/nodenetworkconfigurationpolicies-mutate?timeout=10s\": no endpoints available for service \"nmstate-webhook\`

pinikomarov commented:
/retest

pinikomarov commented:
/retest

raukadah commented:
recheck

raukadah commented:
recheck

pinikomarov commented:
@gibizer I think we can go forward with this patch , people online had had some time to consider, 
I think there are no objections to using dev_mode=true here.
https://redhat-internal.slack.com/archives/CQXJFGMK6/p1689631821764519

pinikomarov commented:
/retest-required

dprince commented on main.go:
```diff
@@ -111,7 +111,7 @@ func main() {
 	devMode, err := strconv.ParseBool(os.Getenv("DEV_MODE"))
 	if err != nil {
-		devMode = false
+		devMode = true
```
this should still default to false be default right?

pinikomarov replied:
I don't think it's necessary , the devMode = true makes the output human readable as in  (top example): 
https://github.com/openstack-k8s-operators/openstack-operator/pull/360#issuecomment-1620038921

and non dev mode is some Json format which is not easy on the eyes.
So I think for debugging purposes and since we don't have yet a CI logging those logs the devmode is the nicer to have option.

pinikomarov commented:
/retest

pinikomarov commented:
/retest

pinikomarov commented:
/retest

pinikomarov commented:
recheck

pinikomarov commented:
recheck

pinikomarov commented:
@dprince @stuggi hi can you review please ?

abays commented:
@gibizer Have your requested changes been satisfied here?

gibizer commented:
> @gibizer Have your requested changes been satisfied here?

Yes, this looks OK to me.

stuggi reviewed (approved):
/lgtm
